src/lib.rs

```diff
@@ -0,0 +1,1039 @@
+//! A concurrent work-stealing deque.
+//!
+//! The data structure can be thought of as a dynamically growable and shrinkable buffer that has
+//! two ends: bottom and top. A [`Deque`] can [`push`] elements into the bottom and [`pop`]
+//! elements from the bottom, but it can only [`steal`][Deque::steal] elements from the top.
+//!
+//! A [`Deque`] doesn't implement `Sync` so it cannot be shared among multiple threads. However, it
+//! can create [`Stealer`]s, and those can be easily cloned, shared, and sent to other threads.
+//! [`Stealer`]s can only [`steal`][Stealer::steal] elements from the top.
+//!
+//! Here's a visualization of the data structure:
+//!
+//! ```text
+//!                    top
+//!                     _
+//!    Deque::steal -> | | <- Stealer::steal
+//!                    | |
+//!                    | |
+//!                    | |
+//! Deque::push/pop -> |_|
+//!
+//!                  bottom
+//! ```
+//!
+//! # Work-stealing schedulers
+//!
+//! Usually, the data structure is used in work-stealing schedulers as follows.
+//!
+//! There is a number of threads. Each thread owns a [`Deque`] and creates a [`Stealer`] that is
+//! shared among all other threads. Alternatively, it creates multiple [`Stealer`]s - one for each
+//! of the other threads.
+//!
+//! Then, all threads are executing in a loop. In the loop, each one attempts to [`pop`] some work
+//! from its own [`Deque`]. But if it is empty, it attempts to [`steal`][Stealer::steal] work from
+//! some other thread instead. When executing work (or being idle), a thread may produce more work,
+//! which gets [`push`]ed into its [`Deque`].
+//!
+//! Of course, there are many variations of this strategy. For example, sometimes it may be
+//! beneficial for a thread to always [`steal`][Deque::steal] work from the top of its deque
+//! instead of calling [`pop`] and taking it from the bottom.
+//!
+//! # Examples
+//!
+//! ```
+//! use crossbeam_deque::{Deque, Steal};
+//! use std::thread;
+//!
+//! let d = Deque::new();
+//! let s = d.stealer();
+//!
+//! d.push('a');
+//! d.push('b');
+//! d.push('c');
+//!
+//! assert_eq!(d.pop(), Some('c'));
```

Comment:
In other data structures, we named to popping function `try_pop()` instead of `pop()`.  I don't have a strong opinion on what's better, but I'd like to have a consistent name.

Reply:
I don't have a strong opinion either... It really depends. Here are some counterexamples:

1. [`Vec::pop`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.pop) (not `try_pop`) returns an `Option<T>`.
2. [`HashMap::remove`](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.remove) (not `try_remove`) returns an `Option<V>`.
3. [`LinkedList::pop_front`](https://doc.rust-lang.org/std/collections/linked_list/struct.LinkedList.html#method.pop_front) (not `try_pop_front`) returns an `Option<T>`.
4. [`BinaryHeap::pop`](https://doc.rust-lang.org/std/collections/binary_heap/struct.BinaryHeap.html#method.pop) (not `try_pop`) returns an `Option<T>`.
5. [`String::pop`](https://doc.rust-lang.org/alloc/string/struct.String.html#method.pop) (not `try_pop`) returns an `Option<char>`.